src/main/java/com/google/jspecify/nullness/NullSpecAnnotatedTypeFactory.java

```diff
@@ -374,6 +375,27 @@ private NullSpecAnnotatedTypeFactory(
     }
   }
 
+  /** Too ensure setRoot is called on both worlds exactly once. */
+  private boolean settingRoot = false;
+
+  /**
+   * Ensure setRoot is called on both worlds exactly once whenever it is called on one of the
+   * worlds.
+   */
+  @Override
+  public void setRoot(@Nullable CompilationUnitTree root) {
+    if (!settingRoot) {
+      settingRoot = true;
+      super.setRoot(root);
+      if (withLeastConvenientWorld != this) {
+        withLeastConvenientWorld.setRoot(root);
+      } else {
+        withMostConvenientWorld.setRoot(root);
+      }
+    }
+    settingRoot = false;
```

Comment:
I might have expected for this to appear inside the `if` so that that block both sets to `true` and sets to `false`. That would also mean that, if someone tried to call `setRoot` _twice_ while a call was already active, then it would still be a no-op.

But I see no reason to think that the current code is a problem.